File: lbcli/src/account.rs
use std::io::Write;

use structopt::StructOpt;

use lb::Core;

use crate::sync;
use crate::util::input;
use crate::CliError;

#[derive(StructOpt)]
pub enum AcctCmd {
    /// Export a private key to stdout.
    Privkey,
    /// Prints out information about your account
    Status,
    /// Create a new subscription using a credit card
    Subscribe,
    /// Terminate a lockbook subscription
    Unsubscribe,
}

#[derive(StructOpt)]
pub struct InitArgs {
    /// Restore a lockbook instance by providing the private key
    #[structopt(long)]
    restore: bool,
    /// Include the welcome doc
    #[structopt(long)]
    welcome: bool,
    /// Don't perform an initial sync
    #[structopt(long)]
    no_sync: bool,
}

pub fn init(core: &Core, args: InitArgs) -> Result<(), CliError> {
    if args.restore {
        if atty::is(atty::Stream::Stdin) {
            return Err(CliError::new("to restore an existing lockbook account, pipe your account string into this command, e.g.:\npbpaste | lockbook init --restore"));
        }

        let mut account_string = String::new();
        std::io::stdin()
            .read_line(&mut account_string)
            .expect("failed to read from stdin");
        account_string.retain(|c| !c.is_whitespace());

        println!("restoring...");
        core.import_account(&account_string)?;

        println!("account restored! next, try syncing.");
    } else {
        print!("please enter a username: ");
        std::io::stdout().flush()?;

        let mut username = String::new();
        std::io::stdin()
            .read_line(&mut username)
            .expect("failed to read from stdin");
        username.retain(|c| c != '\n' && c != '\r');

        let server =
            std::env::var("API_URL").unwrap_or_else(|_| lb::DEFAULT_API_LOCATION.to_string());

        println!("generating keys and checking for username availability...");
        core.create_account(&username, &server, args.welcome)?;

        println!("account created!");
    }
    if !args.no_sync {
        sync(core, false)?;
    }
    Ok(())
}

pub fn acct(core: &Core, cmd: AcctCmd) -> Result<(), CliError> {
    match cmd {
        AcctCmd::Privkey => privkey(core),
        AcctCmd::Status => status(core),
        AcctCmd::Subscribe => subscribe(core),
        AcctCmd::Unsubscribe => cancel_subscription(core),
    }
}

fn privkey(core: &Core) -> Result<(), CliError> {
    let account_string = core.export_account()?;

    let answer: String =
        input("your private key is about to be visible. do you want to proceed? [y/n]: ")?;
    if answer != "y" && answer != "Y" {
        return Ok(());
    }

    println!("{}", account_string);
    Ok(())
}

fn status(core: &Core) -> Result<(), CliError> {
    let cap = core.get_usage()?;
    let pct = (cap.server_usage.exact * 100) / cap.data_cap.exact;

    if let Some(info) = core.get_subscription_info()? {
        match info.payment_platform {
            lb::PaymentPlatform::Stripe { card_last_4_digits } => {
                println!("type: Stripe, *{}", card_last_4_digits)
            }
            lb::PaymentPlatform::GooglePlay { account_state } => {
                println!("type: Google Play");
                println!("state: {:?}", account_state);
            }
        }
        println!("renews on: {}", info.period_end);
    } else {
        println!("trial tier");
    }
    println!("data cap: {}, {}% utilized", cap.data_cap.readable, pct);
    Ok(())
}

fn subscribe(core: &Core) -> Result<(), CliError> {
    println!("checking for existing payment methods...");
    let existing_card =
        core.get_subscription_info()?
            .and_then(|info| match info.payment_platform {
                lb::PaymentPlatform::Stripe { card_last_4_digits } => Some(card_last_4_digits),
                lb::PaymentPlatform::GooglePlay { .. } => None,
            });

    let mut use_old_card = false;
    if let Some(card) = existing_card {
        let answer: String = input(format!("do you want use *{}? [y/n]: ", card))?;
        if answer == "y" || answer == "Y" {
            use_old_card = true;
        }
    } else {
        println!("no existing cards found...");
    }

    let payment_method = if use_old_card {
        lb::PaymentMethod::OldCard
    } else {
        lb::PaymentMethod::NewCard {
            number: input("enter your card number: ")?,
            exp_year: input("expiration year: ")?,
            exp_month: input("expiration month: ")?,
            cvc: input("cvc: ")?,
        }
    };

    core.upgrade_account_stripe(lb::StripeAccountTier::Premium(payment_method))?;
    println!("subscribed!");
    Ok(())
}

fn cancel_subscription(core: &Core) -> Result<(), CliError> {
    println!("cancelling subscription... ");
    core.cancel_subscription()?;
    Ok(())
}

File: lbcli/account.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"os"

	"github.com/steverusso/lockbook-x/go-lockbook"
)

// Account related commands.
//
// clap:cmd_usage <command> [args...]
type acctCmd struct {
	init        *acctInitCmd
	restore     *acctRestoreCmd
	privkey     *acctPrivKeyCmd
	status      *acctStatusCmd
	subscribe   *acctSubscribeCmd
	unsubscribe *acctUnsubscribeCmd
}

func (a *acctCmd) run(core lockbook.Core) error {
	switch {
	case a.init != nil:
		return a.init.run(core)
	case a.restore != nil:
		return a.restore.run(core)
	case a.privkey != nil:
		return a.privkey.run(core)
	case a.status != nil:
		return a.status.run(core)
	case a.subscribe != nil:
		return a.subscribe.run(core)
	case a.unsubscribe != nil:
		return a.unsubscribe.run(core)
	default:
		return nil
	}
}

// Create a lockbook account.
type acctInitCmd struct {
	// Include the welcome document.
	//
	// clap:opt welcome
	welcome bool
	// Don't perform the initial sync.
	//
	// clap:opt no-sync
	noSync bool
}

func (c *acctInitCmd) run(core lockbook.Core) error {
	fi, err := os.Stdin.Stat()
	if err != nil {
		panic(err)
	}
	if fi.Mode()&os.ModeNamedPipe != 0 {
		return errors.New("cannot create a new account without ability to prompt for terminal input")
	}
	scnr := bufio.NewScanner(os.Stdin)
	fmt.Printf("please choose a username: ")
	if !scnr.Scan() {
		return scnr.Err()
	}
	uname := scnr.Text()
	apiURL := os.Getenv("API_URL")
	if apiURL == "" {
		apiURL = lockbook.DefaultAPILocation
	}
	fmt.Println("generating keys and checking for username availability...")
	_, err = core.CreateAccount(uname, c.welcome)
	if err != nil {
		return fmt.Errorf("creating account: %w", err)
	}
	fmt.Println("account created!")
	if !c.noSync {
		fmt.Print("doing initial sync... ")
		if err := core.SyncAll(nil); err != nil {
			return fmt.Errorf("syncing after init: %w", err)
		}
		fmt.Println("done")
	}
	return nil
}

// Restore an existing account from its secret account string.
//
// The restore command reads the secret account string from standard input (stdin).
// In other words, pipe your account string to this command like:
// 'cat lbkey.txt | lbcli restore'.
type acctRestoreCmd struct {
	// Don't perform the initial sync.
	//
	// clap:opt no-sync
	noSync bool
}

func (c *acctRestoreCmd) run(core lockbook.Core) error {
	fi, err := os.Stdin.Stat()
	if err != nil {
		panic(err)
	}
	if fi.Mode()&os.ModeNamedPipe == 0 {
		return errors.New("to restore an existing lockbook account, pipe your account string into this command, e.g.:\npbpaste | lockbook init --restore")
	}
	data, err := io.ReadAll(os.Stdin)
	if err != nil {
		return fmt.Errorf("trying to read from stdin: %w", err)
	}
	if n := len(data); data[n-1] == '\n' {
		data = data[:n-1]
	}
	_, err = core.ImportAccount(string(data))
	if err != nil {
		return fmt.Errorf("importing account: %w", err)
	}
	if !c.noSync {
		fmt.Print("doing initial sync... ")
		if err := core.SyncAll(nil); err != nil {
			return fmt.Errorf("syncing after init: %w", err)
		}
		fmt.Println("done")
	}
	return nil
}

// Print out the private key for this lockbook.
type acctPrivKeyCmd struct{}

func (acctPrivKeyCmd) run(core lockbook.Core) error {
	acctStr, err := core.ExportAccount()
	if err != nil {
		return fmt.Errorf("exporting account: %w", err)
	}
	answer := ""
	fmt.Print("your private key is about to be visible. do you want to proceed? [y/N]: ")
	fmt.Scanln(&answer)
	if answer != "y" && answer != "Y" {
		fmt.Println("aborted")
		return nil
	}
	fmt.Println(acctStr)
	return nil
}

// Overview of your account.
type acctStatusCmd struct{}

func (acctStatusCmd) run(core lockbook.Core) error {
	u, err := core.GetUsage()
	if err != nil {
		return fmt.Errorf("getting usage: %w", err)
	}
	info, err := core.GetSubscriptionInfo()
	if err != nil {
		return fmt.Errorf("getting subscription info: %w", err)
	}
	switch {
	case info.StripeLast4 != "":
		fmt.Printf("type: Stripe, *%s\n", info.StripeLast4)
	case info.GooglePlay != lockbook.GooglePlayNone:
		fmt.Println("type: Google Play")
		fmt.Printf("state: %s\n", info.GooglePlay)
	case info.AppStore != lockbook.AppStoreNone:
		fmt.Println("type: App Store")
		fmt.Printf("state: %s\n", info.AppStore)
	default:
		fmt.Println("trial tier")
	}
	if !info.PeriodEnd.IsZero() {
		fmt.Printf("renews on: %s\n", info.PeriodEnd)
	}
	pct := (u.ServerUsage.Exact * 100) / u.DataCap.Exact
	fmt.Printf("data cap: %s, %d%% utilized\n", u.DataCap.Readable, pct)
	return nil
}

// Create a new subscription with a credit card.
type acctSubscribeCmd struct{}

func (acctSubscribeCmd) run(core lockbook.Core) error {
	fmt.Print("checking for existing card... ")
	subInfo, err := core.GetSubscriptionInfo()
	if err != nil {
		return fmt.Errorf("getting subscription info: %w", err)
	}
	useNewCard := true
	if subInfo.StripeLast4 != "" {
		answer := ""
		fmt.Printf("do you want to use card ending in *%s? [y/N]: ", subInfo.StripeLast4)
		fmt.Scanln(&answer)
		if answer == "y" || answer == "Y" {
			useNewCard = false
		}
	} else {
		fmt.Println("no existing cards found.")
	}
	var card *lockbook.CreditCard
	if useNewCard {
		card := &lockbook.CreditCard{}
		fmt.Print("enter your card number: ")
		fmt.Scanln(&card.Number)
		for {
			fmt.Print("expiration month: ")
			_, err := fmt.Scanln("%d", &card.ExpiryMonth)
			if err == nil {
				break
			}
			fmt.Println(err)
		}
		for {
			fmt.Print("expiration year: ")
			_, err := fmt.Scanf("%d", &card.ExpiryYear)
			if err == nil {
				break
			}
			fmt.Println(err)
		}
		fmt.Print("cvc: ")
		fmt.Scanln(&card.CVC)
	}
	if err = core.UpgradeViaStripe(card); err != nil {
		return err
	}
	fmt.Println("subscribed!")
	return nil
}

// Cancel an existing subscription.
type acctUnsubscribeCmd struct{}

func (acctUnsubscribeCmd) run(core lockbook.Core) error {
	fmt.Print("cancelling subscription... ")
	err := core.CancelSubscription()
	if err != nil {
		return err
	}
	fmt.Println("done")
	return nil
}
